chingu-x/chingu-connect
# Initial setup of graphql with user model

## Purpose
* Implement graphql on backend

## Approach/Fixes
* Setup basic graphql config
  * Created `RootQuery`
  * Created `UserType`
  * Created `mutations`
    * Created an `addUser` mutations
* Re-factored db to have `models` and `schema` directories
* Re-factored `user.js` to use `mongoose.model` instead of implementing { model } from mongoose
  * Syntax doesn't seem to be supported yet
  * Reference: https://github.com/Automattic/mongoose/issues/6379

## Testing
* Download this PR or branch to your local device
* Run `npm install` to grab new dependencies
* Run `npm run dev` to boot up the app 
  * If using local mongodb, boot up your server first
    * Run `mongod` for mac or `mongod.exe` for windows (I think?)
* In browser, go to localhost:8008/graphql to see if everything is setup properly
  * Run the following code to see users list (there shouldn't be any to start I think?):
```
{
  users {
    id
  }
}
```
  * Run the following code to add a new user:
```
mutation {
  addUser (username: "<enter_username_here>", email: "<enter_email_here>") {
    username
    email
  }
}
```
  * Run the following code to see newly created user:
```
{
  user(id: "<newly_created_id">) {
    username
    email
  }
}
```

## Notes
* This is my first time setting up graphql so I may have missed stuff
* I'm also currently using express-graphql instead of apollo-server, but we can switch if need be (this is just how I know to set it up based on the tutorials I followed earlier)
* I wanted to implement just the basic setup so that we can re-factor/add to it afterwards
